scipy/special/special/tools.h

```diff
@@ -73,5 +75,273 @@ namespace detail {
+        return {f, 0};
+    }
+
+    /* Performs one step of Kahan summation. */
+    template <typename T>
+    void kahan_step(T *sum, T *comp, T x) {
+        T y = x - *comp;
+        T t = *sum + y;
+        *comp = (t - *sum) - y;
+        *sum = t;
+    }
+
+    /* Evaluates a continued fraction using the "series" method.
+     *
+     * Denote the continued fraction by
+     *
+     *              a[1]   a[2]   a[3]
+     *   f = b[0] + ------ ------ ------ ...
+     *              b[1] + b[2] + b[3] +
+     *
+     * Denote the n-th convergent by f[n], starting from f[0] := b[0].  The
+     * series method (see [1])
+     *
+     * Reference
+     * ---------
+     *
+     * [1] Gautschi, W. (1967). “Computational Aspects of Three-Term
+     *     Recurrence Relations.” SIAM Review, 9(1):24-82.
+     *
+     * Parameters
+     * ----------
+     *
+     *   init_val
+     *       b0.  The type of this parameter (T) is used for intermediary
+     *       computations as well as the return value.  It is expected to
+     *       be a (possibly complex) floating point type.
+     *
+     *   cf
+     *       Input iterator that yields the terms of the continued fraction.
+     *       It must support two operations: `*cf` returns the current term
+     *       (a[n], b[n]) as a pair; `++cf` advances to the next term. The
+     *       iterator initially points to (a[1], b[1]).
+     *
+     *   tol
+     *       Tolerance used to terminate the iteration.  Specifically, stop
+     *       iteration as soon as `abs(f[n] - f[n-1]) <= abs(tol * f[n])`,
+     *       where f[n], n >= 1 is the n-th convergent, and f[0] = b0.  The
+     *       value of `tol` should be chosen to bound the truncation error.
+     *
+     *   max_terms
+     *       Maximum number of terms to evaluate.  This parameter is used to
+     *       handle unexpected non-convergence.  It should normally be set
+     *       large enough such that the continued fraction is mathematically
+     *       guaranteed to have converged within that many terms.
+     *
+     * Return Value
+     * ------------
+     *
+     * If the continued fraction converges (i.e. the tolerance condition is
+     * satisfied) by f[n] with `n <= max_terms`, returns (f[n], n).
+     * Otherwise, returns (f[max_terms], 0).
+     *
+     * Remarks
+     * -------
+     *
+     * This is a low-level routine intended for expert use.  No error checking
+     * is performed.  The caller must ensure that the parameters and terms are
+     * finite and that intermediary calculations do not trigger floating point
+     * exceptions such as overflow.
+     *
+     * The numerical accuracy of this method depends on the characteristics of
+     * the continued fraction being evaluated.
+     */
+    template <typename T, typename InputIt>
+    SPECFUN_HOST_DEVICE std::pair<T, std::size_t> continued_fraction_eval_series(
+        T init_val, InputIt cf, T tol, std::size_t max_terms) {
```

Comment:
Can we factor this so that we define a new iterator which takes an iterator for a continued fraction and tranforms it into an iterator for the terms of thid series, which can then be fed into a series evaluator, perhaps a regular one, a Kahan summation evaluator. We also want to add convergence acceleration transforms that series iterators could be passed into.

Reply:
Sounds good. I can make something like `ContinuedFractionDifferenceSeries` to yield the adjacent difference between partial convergents. That will make the logic clearer. As for "without worrying about details", I'm afraid the caller still need to do some serious analysis. For example, a series evaluator can be as good as it is, but if the caller supplies a slow converging series, they can't expect it to run quickly!

Reply:
Cool. Understanding the math is needed, but I want people who know the math to be able to write code at a higher , almost declarative level, without needing to worry too much about programming details. `scipy.special` was unmaintained for long time because it requires knowledge both of the math and programming in compiled languages, and the hope is to simplify things for people who are stronger in the former.